File: demos/go/bytebox.go
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2025 Contributors to the bytebox-wasm project
 *
 * Part of the bytebox-wasm project - https://github.com/danielspk/bytebox-wasm
 * See LICENSE file for full license details.
 */

package main

// ----------------------------------------------------------------------------
// Console Constants
// ----------------------------------------------------------------------------

const (
	SCREEN_WIDTH     uint32 = 160
	SCREEN_HEIGHT    uint32 = 144
	FRAMEBUFFER_SIZE uint32 = (SCREEN_WIDTH * SCREEN_HEIGHT / 4)
)

// ----------------------------------------------------------------------------
// Memory Addresses Constants
// ----------------------------------------------------------------------------

const (
	SYSFLAGS_ADDR  uint32 = 0x0040
	SEED_ADDR	   uint32 = 0x0041
	GAMENAME_ADDR  uint32 = 0x0044
	VIDEO_ADDR     uint32 = 0xE900
	COLOR1_ADDR    uint32 = 0xFF84
	COLOR2_ADDR    uint32 = 0xFF87
	COLOR3_ADDR    uint32 = 0xFF8A
	COLOR4_ADDR    uint32 = 0xFF8D
	GAMEPAD1_ADDR  uint32 = 0xFF94
	GAMEPAD2_ADDR  uint32 = 0xFF95
	SFX_CH1_ADDR   uint32 = 0xFF98
	SFX_CH2_ADDR   uint32 = 0xFF9C
	SFX_CH3_ADDR   uint32 = 0xFFA0
	SFX_CH4_ADDR   uint32 = 0xFFA4
)

// ----------------------------------------------------------------------------
// System Flags Bits Constants
// ----------------------------------------------------------------------------

const HALT_FLAG uint32 = 0x01
const DUMP_WRAM_FLAG uint32 = 0x02

// ----------------------------------------------------------------------------
// Gamepad Buttons Bits Constants
// ----------------------------------------------------------------------------

const (
	BUTTON_2     uint32 = 0x01
	BUTTON_1     uint32 = 0x02
	BUTTON_RIGHT uint32 = 0x10
	BUTTON_DOWN  uint32 = 0x20
	BUTTON_UP    uint32 = 0x40
	BUTTON_LEFT  uint32 = 0x80
)

// ----------------------------------------------------------------------------
// Imported functions
// ----------------------------------------------------------------------------

// Gets the value of a memory address
//
//go:wasmimport env peek
func peek(addr uint32) uint32

// Sets a value into a memory address
//
//go:wasmimport env poke
func poke(addr uint32, value uint32)

// Puts trace information to the console
//
//go:wasmimport env trace
func trace(msg string, len uint32)

File: demos/templates/assemblyscript/bytebox.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2025-2026 Contributors to the bytebox-wasm project
//
// Part of the bytebox-wasm project - https://github.com/danielspk/bytebox-wasm
// See LICENSE file for full license details.

// ----------------------------------------------------------------------------
// Console Constants
// ----------------------------------------------------------------------------

export const SCREEN_WIDTH: u16 = 160;
export const SCREEN_HEIGHT: u16 = 120;
export const FRAMEBUFFER_SIZE: u32 = <u32>SCREEN_WIDTH * <u32>SCREEN_HEIGHT / 4;

// ----------------------------------------------------------------------------
// Memory Addresses Constants
// ----------------------------------------------------------------------------

export const SYSFLAGS_ADDR: u16 = 0x0040;
export const SEED_ADDR: u16 = 0x0041;
export const GAMENAME_ADDR: u16 = 0x0044;
export const WRAM_ADDR: u16 = 0xE100;
export const VIDEO_ADDR: u16 = 0xE900;
export const COLOR1_ADDR: u16 = 0xFF84;
export const COLOR2_ADDR: u16 = 0xFF87;
export const COLOR3_ADDR: u16 = 0xFF8A;
export const COLOR4_ADDR: u16 = 0xFF8D;
export const GAMEPAD1_ADDR: u16 = 0xFF94;
export const GAMEPAD2_ADDR: u16 = 0xFF95;
export const SFX_CH1_ADDR: u16 = 0xFF98;
export const SFX_CH2_ADDR: u16 = 0xFF9C;
export const SFX_CH3_ADDR: u16 = 0xFFA0;
export const SFX_CH4_ADDR: u16 = 0xFFA4;
export const MELODY_ATTR_ADDR: u16 = 0xFFA8;
export const MELODY_ADDR: u16 = 0xFFA9;
export const MELODY_HEAD_ADDR: u16 = 0xFFE9;
export const MELODY_TAIL_ADDR: u16 = 0xFFEA;

// ----------------------------------------------------------------------------
// System Flags Bits Constants
// ----------------------------------------------------------------------------

export const HALT_FLAG: u8 = 0x01;
export const DUMP_WRAM_FLAG: u8 = 0x02;

// ----------------------------------------------------------------------------
// Gamepad Buttons Bits Constants
// ----------------------------------------------------------------------------

export const BUTTON_2: u8 = 0x01;
export const BUTTON_1: u8 = 0x02;
export const BUTTON_RIGHT: u8 = 0x10;
export const BUTTON_DOWN: u8 = 0x20;
export const BUTTON_UP: u8 = 0x40;
export const BUTTON_LEFT: u8 = 0x80;

// ----------------------------------------------------------------------------
// Imported functions
// ----------------------------------------------------------------------------

@external("env", "peek")
export declare function peek(addr: u16): u8;

@external("env", "poke")
export declare function poke(addr: u16, value: u8): void;

@external("env", "spoke")
export declare function spoke(start_addr: u16, length: u16, data_ptr: usize): void;

@external("env", "trace")
export declare function trace(str_ptr: usize, len: i32): void;